### DISCUSSION TITLE
How to remove an event listener before a route changes?

### DISCUSSION THREAD

**Discussion by IcyFoxe** [vercel/next.js#19952] [Help] [answered]
When I register an event listener in the "useEffect" function and then remove it in the return statement, it removes this event listener AFTER the route changes, which causes warnings about memory leaks in React.

Is there a solution to that?

Example:
```
useEffect(() => {
  window.addEventListener("scroll", scrollFunction)
  return () => {
    window.removeEventListener("scroll", scrollFunction) // this event listener is removed after the new route loads
  }
}, [])
```

**Edit: It removes the event listener as it should, I just had setTimeout() running that leaked over to the next page..**

**Comment by IcyFoxe** [comment, marked as answer]
So my current solution to the problem is using NextJS useRouter hook.

I use the `router.events.on('routeChangeStart', removeEventListenersFunction)` event to remove my event listeners before the page changes. It seems like somewhat of a hacky solution though..

Is there a better way to solve this?

**Edit:** I just realized that I had setTimeout() running that was leaking over to the next page, so that's what was the problem.

**Reply by nghiepnguyen520** [reply]
Are you sure this will auto remove when I change pages?

**Reply by IcyFoxe** [reply]
No, it doesn't. Ignore what I said in the comment before the "edit" sentence..
You have to remove all event listeners and timeouts in the useEffect return function, otherwise they will cause problems.
